### PR TITLE
Recommend using the cuda-nvcc package from the "nvidia" conda channel…

### PR DESCRIPTION
**Pull request by hawkinsp**
… to fetch ptxas.

@ngam

### CODE REVIEW

**Review by ngam — CHANGES_REQUESTED** [review]
Actually, let's do this properly. (Bear with me, sorry about the verbosity!) 

~First, I will keep an eye on what the nvidia channel publishes. The only potential problem would be if they publish go back to publishing cudatoolkit, cudnn, and nccl themselves more regularly, then we may end up in situations where these would come from a mix of channels which could confuse things. Even though, I would generally think it would be okay because we simply repackage their stuff anyway, so it is not like we are building things from scratch or anything.~ Obviously, this is the problem below... Now you get why we don't like to mix channels!

~Second, l~ Let's look at a few scenarios:


for reference, either `conda create -n test_cuda_nvda jax cuda-nvcc -c nvidia -c conda-forge` or `CONDA_OVERRIDE_CUDA="11.2" conda create -n test_cuda_nvda jax cuda-nvcc -c nvidia -c conda-forge` yields

```
 _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge
  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-2_gnu
  absl-py            conda-forge/noarch::absl-py-1.2.0-pyhd8ed1ab_0
  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4
  c-ares             conda-forge/linux-64::c-ares-1.18.1-h7f98852_0
  ca-certificates    conda-forge/linux-64::ca-certificates-2022.6.15-ha878542_0
  cuda-nvcc          nvidia/linux-64::cuda-nvcc-11.7.99-0
  etils              conda-forge/noarch::etils-0.7.1-pyhd8ed1ab_0
  grpc-cpp           conda-forge/linux-64::grpc-cpp-1.46.4-hc2bec63_2
  importlib_resourc~ conda-forge/noarch::importlib_resources-5.9.0-pyhd8ed1ab_0
  jax                conda-forge/noarch::jax-0.3.16-pyhd8ed1ab_0
  jaxlib             conda-forge/linux-64::jaxlib-0.3.15-cpu_py310h3dca822_0
  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.36.1-hea4e1c9_2
  libabseil          conda-forge/linux-64::libabseil-20220623.0-cxx17_h48a1fff_1
  libblas            conda-forge/linux-64::libblas-3.9.0-16_linux64_openblas
  libcblas           conda-forge/linux-64::libcblas-3.9.0-16_linux64_openblas
  libffi             conda-forge/linux-64::libffi-3.4.2-h7f98852_5
  libgcc-ng          conda-forge/linux-64::libgcc-ng-12.1.0-h8d9b700_16
  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-12.1.0-h69a702a_16
  libgfortran5       conda-forge/linux-64::libgfortran5-12.1.0-hdcd56e2_16
  libgomp            conda-forge/linux-64::libgomp-12.1.0-h8d9b700_16
  liblapack          conda-forge/linux-64::liblapack-3.9.0-16_linux64_openblas
  libnsl             conda-forge/linux-64::libnsl-2.0.0-h7f98852_0
  libopenblas        conda-forge/linux-64::libopenblas-0.3.21-pthreads_h78a6416_0
  libprotobuf        conda-forge/linux-64::libprotobuf-3.20.1-h6239696_1
  libsqlite          conda-forge/linux-64::libsqlite-3.39.2-h753d276_1
  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-12.1.0-ha89aaad_16
  libuuid            conda-forge/linux-64::libuuid-2.32.1-h7f98852_1000
  libzlib            conda-forge/linux-64::libzlib-1.2.12-h166bdaf_2
  ncurses            conda-forge/linux-64::ncurses-6.3-h27087fc_1
  numpy              conda-forge/linux-64::numpy-1.23.1-py310h53a5b5f_0
  openssl            conda-forge/linux-64::openssl-3.0.5-h166bdaf_1
  opt_einsum         conda-forge/noarch::opt_einsum-3.3.0-pyhd8ed1ab_1
  pip                conda-forge/noarch::pip-22.2.2-pyhd8ed1ab_0
  python             conda-forge/linux-64::python-3.10.5-ha86cf86_0_cpython
  python_abi         conda-forge/linux-64::python_abi-3.10-2_cp310
  re2                conda-forge/linux-64::re2-2022.06.01-h27087fc_0
  readline           conda-forge/linux-64::readline-8.1.2-h0f457ee_0
  scipy              conda-forge/linux-64::scipy-1.9.0-py310hdfbd76f_0
  setuptools         conda-forge/linux-64::setuptools-65.0.1-py310hff52083_0
  sqlite             conda-forge/linux-64::sqlite-3.39.2-h4ff8645_1
  tk                 conda-forge/linux-64::tk-8.6.12-h27826a3_0
  typing_extensions  conda-forge/noarch::typing_extensions-4.3.0-pyha770c72_0
  tzdata             conda-forge/noarch::tzdata-2022c-h191b570_0
  wheel              conda-forge/noarch::wheel-0.37.1-pyhd8ed1ab_0
  xz                 conda-forge/linux-64::xz-5.2.6-h166bdaf_0
  zipp               conda-forge/noarch::zipp-3.8.1-pyhd8ed1ab_0
  zlib               conda-forge/linux-64::zlib-1.2.12-h166bdaf_2

```

Note the "cpu" version. This is quite annoying and I need to figure out how to make it pull the gpu jaxlib when applicable.

Even adding `jaxlib` explicitly doesn't resolve this, e.g. `CONDA_OVERRIDE_CUDA="11.2" conda create -n test_cuda_nvda jax jaxlib cuda-nvcc -c nvidia -c conda-forge` yields the same thing above.

Now, let's be even more explicit --- let's ask for the cuda builds directly. Unfortunately, this `mamba create -n test_cuda_nvda jax jaxlib==*=*cuda* cuda-nvcc -c conda-forge` errors with 

```
Encountered problems while solving:
  - package jaxlib-0.3.14-cuda112py310h0a0fd6d_203 requires cudnn >=8.4.1.50,<9.0a0, but none of the providers can be installed
```

(sorry I switched to mamba because it is faster to arrive at the error, but the same should apply to conda)

Finally, this resolves things correctly:

```
~$ mamba create -n test_cuda_nvda jax jaxlib==*=*cuda* cuda-nvcc -c conda-forge -c nvidia
```

TLDR: channel order matters, so `-c conda-forge` should come before `-c nvidia` ~and we need to explicitly ask for the jaxlib cuda version.~
- If `-c nvidia` is first, it will start looking for cudnn, cudatoolkit, etc. from the nvidia channel first and if the user has `--strict-channel` policy it will fail.
- ~I will work on fixing this so that the cuda version could be requested directly from jax. I will need to think about a way to do this...  I am surprised it is not happening automatically actually.~ this is fine actually if the channel order is correct.

File: README.md [hunk 1]
```diff
@@ -518,14 +518,18 @@ There is a community-supported Conda build of `jax`. To install using `conda`,
...
 ```
 
-If run on a machine with NVidia GPUs, this will automatically install a
-GPU-enabled `jaxlib`. The conda-forge project redistributes CUDA and CuDNN; as
-such, there is no need to install CuDNN beforehand. However, the `cudatoolkit`
-distributed by `conda-forge` is missing `ptxas`, which JAX requires. You must
-therefore install CUDA on your machine yourself so that `ptxas` is in your path.
+To install on a machine with an NVidia GPU, run
+```bash
+conda install jax cuda-nvcc -c nvidia -c conda-forge
```
[the remark this answers is not in the document]
Comment (hawkinsp) [reply README.md]: Done.

File: README.md [hunk 2]
```diff
@@ -518,14 +518,18 @@ There is a community-supported Conda build of `jax`. To install using `conda`,
...
+```
+
+Note the `cudatoolkit` distributed by `conda-forge` is missing `ptxas`, which
+JAX requires. You must therefore either install the `cuda-nvcc` package from
+NVIDIA, or install CUDA on your machine separately so that `ptxas` is in your
+path.
```
[the remark this answers is not in the document]
Comment (hawkinsp) [reply README.md:534]: Done, reworded a bit.

**Review by ngam — APPROVED** [review]
Thanks a ton for taking the time to document and dig into the weeds of this! You're teaching me about conda at this point!

**Comment by jaimergp** [comment]
```
CONDA_OVERRIDE_CUDA="11.2" conda create -n test_cuda_nvda jax jaxlib nvidia::cuda-nvcc -c conda-forge
```

This syntax might work better @ngam
